segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLogFactory.java

```diff
@@ -169,5 +189,84 @@ private BookKeeper startBookKeeperClient() throws Exception {
                          .build();
     }
 
+    /**
+     * Recreate the Bookkeeper client if a given log exhibits MAX_CREATE_ATTEMPTS_PER_LOG creation attempts (as a proxy
+     * for Container recoveries) within the period of time defined in LOG_CREATION_INSPECTION_PERIOD.
+     *
+     * @param logId Id of the log being restarted.
+     */
+    private void tryResetBookkeeperClient(int logId) {
+        synchronized (this) {
+            LogInitializationRecord record = logInitializationTracker.get(logId);
+            if (record != null) {
+                // Account for a restart of the Bookkeeper log.
+                record.incrementLogCreations();
+                // If the number of restarts for a single container is meets the threshold, let's reset the BK client.
+                if (record.isBookkeeperClientResetNeeded()
+                        && lastBookkeeperClientReset.get().getElapsed().compareTo(LOG_CREATION_INSPECTION_PERIOD) > 0) {
+                    try {
+                        log.info("Start creating Bookkeeper client in reset.");
+                        BookKeeper newClient = startBookKeeperClient();
+                        // If we have been able to create a new client successfully, reset the current one and update timer.
+                        log.info("Successfully created new Bookkeeper client, setting it as the new one to use.");
+                        BookKeeper oldClient = this.bookKeeper.getAndSet(newClient);
+                        lastBookkeeperClientReset.set(new Timer());
+                        // Lastly, attempt to close the old client.
+                        log.info("Attempting to close old client.");
+                        oldClient.close();
```

Comment:
this "close()" will fail all of the pending operations 
IIUC with this change we are going to close the BK client even if it is not needed

Reply:
@eolivelli correct, this defensive mechanism is going to reset the Bookkeeper client (shared across all containers) if we detect X number of restarts for a given container within a period of time T. I understand that this is not the ideal way to handle this problem, but this may be a temporarily solution to the problem reported in the issue until this is fixed: https://github.com/apache/bookkeeper/issues/2482. In parallel, we are working on another solution related to better set the `TCP_USER_TIMEOUT` as suggested by @fpj. If that works, maybe we do not need to merge this PR.

Reply:
I see, thanks for your clarification